stan/math/prim/mat/fun/append_array.hpp

```diff
@@ -0,0 +1,89 @@
+
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/scal/meta/is_vector_like.hpp>
+#include <stan/math/prim/scal/meta/return_type.hpp>
+#include <boost/math/tools/promotion.hpp>
+#include <boost/utility/enable_if.hpp>
+#include <vector>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * Return the concatenation of two specified integer vectors in the
+     *   order of the arguments.
+     *
+     * @param x First vector
+     * @param y Second vector
+     * @return A vector of x and y concatenated together (in that order)
+     */
+    inline std::vector<int>
+    append_array(const std::vector<int>& x, const std::vector<int>& y) {
+      std::vector<int> z;
+      z.reserve(x.size() + y.size());
+      z.insert(z.end(), x.begin(), x.end());
+      z.insert(z.end(), y.begin(), y.end());
+      return z;
+    }
+
+    /**
+     * Return the concatenation of two vectors containing Matrices
+     *   of the same type in the order of the arguments.
+     *
+     * @tparam T1 Type of element in Matrices in first vector
+     * @tparam T2 Type of element in Matrices in second vector
+     * @tparam R Row specification of matrices
+     * @tparam C Column specification of matrices
+     * @param x First vector
+     * @param y Second vector
+     * @return A vector of x and y concatenated together (in that order)
+     */
+    template <typename T1, typename T2, int R, int C>
+    inline std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> >
+    append_array(const std::vector<Eigen::Matrix<T1, R, C> >& x,
+                 const std::vector<Eigen::Matrix<T2, R, C> >& y) {
+      if (!x.empty() && !y.empty()) {
+        check_matching_dims("append_array",
+                            "x[1]",
+                            x.front(),
+                            "y[1]",
+                            y.front());
+      }
+
+      std::vector<Eigen::Matrix<typename return_type<T1, T2>::type, R, C> > z;
+      z.reserve(x.size() + y.size());
+      for (size_t i = 0; i < x.size(); i++)
+        z.push_back(x[i].template cast<typename return_type<T1, T2>::type>());
+      for (size_t i = 0; i < y.size(); i++)
+        z.push_back(y[i].template cast<typename return_type<T1, T2>::type>());
+      return z;
+    }
+
+    /**
+     * Return the concatenation of two specified vectors in the order of
+     *   the arguments. The types in each vector must not be vector-like
+     *   themselves.
+     *
+     * @tparam T1 Scalar type of first vector
+     * @tparam T2 Scalar Type of second vector
+     * @param x First vector
+     * @param y Second vector
+     * @return A vector of x and y concatenated together (in that order)
+     */
+    template <typename T1, typename T2>
+    inline typename
+    boost::disable_if_c<is_vector_like<T1>::value || is_vector_like<T2>::value,
+      std::vector<typename
+                  boost::math::tools::promote_args<T1, T2>::type> >::type
```

Comment:
Why do you use `promote_args` here and `return_type` on the implementation above (line 44)?